### PR TITLE
Fix camera simulator

### PR DESCRIPTION
**Pull request by AnthonyHorton**
Fix for bugs discussed in https://github.com/panoptes/POCS/issues/478

### CODE REVIEW

**Comment by wtgee** [comment]
I'm going to pull this down and run it to see if I can find the same messages in the logs that I was experiencing the other day. Will approve after that. Thanks for getting in!

**Comment by wtgee** [comment]
Wow, coverage up by 2.1%. Yes, I think this was causing errors in the code but not failing the tests, so we must now be getting further on in tests. 👍

**Comment by wtgee** [comment]
Ok, I think I see the issue and it has to do with a bad naming choice on my part. In the `_fake_exposure` method you are grabbing the `unsolved.fits ([here](https://github.com/panoptes/POCS/blob/develop/pocs/camera/simulator.py#L100)), but that it is a fits file that should really be named `unsolvable.fits` as it is the one we use to test what happens when solving fails. So when I run pocs with all simulators (via pocs_shell or jupyter) it will fail on the solve in the `pointing` state and so shut itself down. 

There is a valid question to be asked of whether or not we should shut down if solving fails or not, but I think for here and now we should change the file to `solved.fits` that it grabs data from as then solving will continue. @anthonyhorton, is there any reason you would want those fake exposures to not be solving?

**Comment by AnthonyHorton** [comment]
@wtgee No, I don't have any reasons to want the solves to fail. I'll change it so it copies the data from `solved.fits` instead.

**Comment by wtgee** [comment]
There's still actually a couple of things going on:

* The headers are not copied but are generated from the current observation yet saved with the data from `solved.fits`. This is then suggesting an RA/Dec to solve-field that doesn't match the data, which is causing it to timeout without solving (which takes ~30 seconds and may be contributing partially to #191).
* Simulator images are saved in `$PANDIR/images/fields` under the camera simulator name but are (sometimes?) not properly cleaned up.

**Comment by AnthonyHorton** [comment]
Ah, I see. Yes, the simulator doesn't copy headers from `solved.fits` because I wanted the simulator to use as much of the real camera code as possible, including header generation. I could just overwrite the mount coordinates values with the ones from

**Comment by wtgee** [comment]
I tried to just override the mount coordinates in the header but didn't seem to help. I'll do a little more exploring and then report back.

Also not entirely sure what to do about cleanup. I think it's better to have them going to the same location but it's a lot of management to clean up. Not terrible during tests but it means anyone using a simulator (e.g. manually) has to know to clean them up. Could be handled as part of `Observatory.cleanup_observations_dir` but that requires running that too. I'll give it a little more thought.

**Comment by wtgee** [comment]
Ok, one problem which might explain the intermittent behavior is the name of the camera simulator. You are assigning a random int [here](/var/panoptes/images/fields/Wasp2/SC9593/20180213T221404//pointing00.fits) but the format string doesn't have leading zeros, so if number is less than 1000 it ends up with a space in the name, which breaks things. Changing formatting to `{:04d}` solves this issue.

**Comment by wtgee** [comment]
I ended up overwriting the coords in `_fake_exposure` with:

```
solved_header = fits.getheader(file_path)
header['RA-MNT'] = solved_header['RA-MNT']
header['HA-MNT'] = solved_header['HA-MNT']
header['DEC-MNT'] = solved_header['DEC-MNT']
``` 

And then adding a copy of `_process_fits` to the simulator but remove the call to `update_headers`. Then things appear to be working ok.

**Comment by wtgee** [comment]
Here's a patch if you want.

[patch.txt](https://github.com/panoptes/POCS/files/1722011/patch.txt)

**Comment by wtgee** [comment]
One other change, the `simulator.take_observation` needs an `is_primary` for adding to exposure list:

```
        # Add most recent exposure to list
        if self.is_primary:
            observation.exposure_list[image_id] = file_path
```

**Comment by AnthonyHorton** [comment]
Ah, yes, the mount coordinates are not normally added until `process_exposure()` calls `_process_fits()`, which calls `update_headers()`, so any values added in `_fake_exposure` will get overwritten.

As an alternative to overriding `process_fits()` to remove the `update_headers()` call I could instead call `update_headers()` then overwrite the mount coordinates.  Trying to exercise as much of the real camera code as possible with the simulator to keep the test coverage up.

**Comment by AnthonyHorton** [comment]
One thing confusing me, when & how does the field solving get triggered?

**Comment by AnthonyHorton** [comment]
OK, I _think_ this should work now. With the addition of the `is_primary` check to `take_observation()` it became almost identical to the base class `take_observation` method. The only difference is the exposure time trimming so now I just do that then call `super().take_observation()`.

I've also added an override for `_process_fits()` that calls `super()._process_fits()` then overwrites with mount coordinates header entries with those from `solved.fits`.

I also took the opportunity to update a docstring and improve the logic/consistency of a couple of variable names in `camera.py`.

File: pocs/camera/simulator.py [hunk 1]
```diff
@@ -107,3 +93,14 @@ def _fake_exposure(self, filename, header, exposure_event):
                                           dtype=fake_data.dtype)
 
         fits_utils.write_fits(fake_data, header, filename, self.logger, exposure_event)
+
+    def _process_fits(sefl, file_path, info):
```
Comment (wtgee) [inline pocs/camera/simulator.py]: `sefl` -> `self`

File: pocs/camera/camera.py [hunk 2]
```diff
@@ -224,7 +226,7 @@ def process_exposure(self, info, signal_event, exposure_event=None):
 
         Args:
             info (dict): Header metadata saved for the image
-            signal_event (threading.Event): An event that is set signifying that the
+            observation_event (threading.Event): An event that is set signifying that the
```
Comment (wtgee) [inline pocs/camera/camera.py]: Line 271 needs to get changed to match. Won't let me comment on it here.

File: pocs/camera/simulator.py [hunk 3]
```diff
@@ -107,3 +93,14 @@ def _fake_exposure(self, filename, header, exposure_event):
...
+        super()._process_fits(file_path, info)
+        self.logger.debug('Overriding mount coordinates for camera simulator')
+        solved_path = "{}/pocs/tests/data/{}".format(os.getenv('POCS'), 'solved.fits')
+        solved_header = fits.getheader(solved_path)
+        with fits.open(file_path, 'update') as f:
+            hdu = f[0]
+            hdu.header.set('RA-MNT', solved_header['RA-MNT'], 'Degrees')
+            hdu.header.set('HA-MNT', solved_header['HA-MNT'], 'Degrees')
+            hdu.header.set('DEC-MNT', solved_header['DEC-MNT'], 'Degrees')
```
Comment (wtgee) [inline pocs/camera/simulator.py]: Need to `return file_path`

File: pocs/camera/simulator.py [hunk 4]
```diff
@@ -26,37 +26,23 @@ def connect(self):
...
     def take_observation(self, observation, headers=None, filename=None, *args, **kwargs):
-        camera_event = Event()
 
-        exp_time, file_path, image_id, metadata = self._setup_observation(observation,
-                                                                          headers,
-                                                                          filename,
-                                                                          *args,
-                                                                          **kwargs)
+        exp_time = kwargs.get('exp_time', observation.exp_time.value)
+        if exp_time > 5:
```
Comment (wtgee) [inline pocs/camera/simulator.py]: Can you change this to `2` while you are at it? I just tested and works fine.

**Comment by AnthonyHorton** [comment]
+2.62%!

**Comment by wtgee** [comment]
Awesome. Looks good for merging to me. Anything else you wanted in there?

**Comment by AnthonyHorton** [comment]
No, pretty happy with it now.

**Pull request merged**